src/check/check_types.zig

```diff
@@ -267,15 +267,90 @@ pub fn checkExpr(self: *Self, expr_idx: CIR.Expr.Idx) void {
+
+                // Unify each element's var with the list's elem var
+                const result = self.unify(elem_var, @enumFromInt(@intFromEnum(elem_expr_id)));
+
+                switch (result) {
+                    .ok => {},
+                    .problem => |problem_idx| {
+                        // Unification failed, so we know it appended a type mismatch to self.problems.
+                        // We'll translate that generic type mismatch between the two elements into
+                        // a more helpful list-specific error report.
+
+                        // Extract info from the type mismatch problem
+                        var elem_var_snapshot: snapshot.SnapshotContentIdx = undefined;
+                        var incompatible_snapshot: snapshot.SnapshotContentIdx = undefined;
+
+                        // Extract snapshots from the type mismatch problem
+                        switch (self.problems.problems.get(problem_idx)) {
+                            .type_mismatch => |mismatch| {
+                                // The expected type is elem_var, actual is the incompatible element
+                                elem_var_snapshot = mismatch.expected;
+                                incompatible_snapshot = mismatch.actual;
+
+                                // Include the previous element in the error message, since it's the one
+                                // that the current element failed to unify with.
+                                const prev_elem_expr = self.can_ir.store.getExpr(last_unified_idx);
+                                const incompatible_elem_expr = self.can_ir.store.getExpr(elem_expr_id);
+                                const prev_region = prev_elem_expr.toRegion();
+                                const incomp_region = incompatible_elem_expr.toRegion();
+
+                                // Replace the generic Problem in the MultiArrayList with a list-specific one
+                                self.problems.problems.set(problem_idx, .{
+                                    .incompatible_list_elements = .{
+                                        .list_region = list.region,
+                                        .first_elem_region = prev_region orelse list.region,
+                                        .first_elem_var = @enumFromInt(@intFromEnum(last_unified_idx)),
+                                        .first_elem_snapshot = elem_var_snapshot,
+                                        .first_elem_index = last_unified_index,
+                                        .incompatible_elem_region = incomp_region orelse list.region,
+                                        .incompatible_elem_var = @enumFromInt(@intFromEnum(elem_expr_id)),
+                                        .incompatible_elem_snapshot = incompatible_snapshot,
+                                        .incompatible_elem_index = i,
+                                        .list_length = elems.len,
```

Comment:
We might be able to save a little bit more memory here in `problems` by not writing down the regions. Since we have the `Var`s captured, and `problems.buildIncompatibleListElementsReport` is passed `*const can.CIR` as an argument, we could lookup the region at the time of error generation by converting the `Var` to a `CIR.Node.Idx`